File: java/amp-rabbit/src/test/java/amp/rabbit/topology/ExchangeTests.java
package amp.rabbit.topology;

import java.util.HashMap;

import org.junit.*;

import static org.junit.Assert.*;

/** 
*  These tests assert that equality and hashcode generation use value semantics
*  not just in-so-far as properties have the same reference values but that 
* properties themselves are compared using a value semantic.  
*/
public class ExchangeTests extends AmqpBaseModelTests{

	protected Exchange _typeNotEquivelentModel = new Exchange();
	
	@Override
	public void setup(){
		
        _referenceModel = NewExchange("topic", "arg1");
        _equivelentModel = NewExchange("topic", "arg1");
        _notEquivelentModel = NewExchange("topic", "arg2");
        _typeNotEquivelentModel = NewExchange("direct", "arg1");

	}
	
    private Exchange NewExchange(String exchangeType, String argValue)
    {
    	Exchange exchange = new Exchange();
        exchange.arguments = new HashMap<String,Object>();
        exchange.arguments.put("1", argValue);
        exchange.setExchangeType(exchangeType);
        return exchange;
    }
    
	@Test
	public void Equality_should_consider_exchange_type(){
		assertFalse(_referenceModel.equals(_typeNotEquivelentModel));
	}

	@Test
	public void Hashcodes_should_differ_if_exchange_type_differ(){
		assertNotEquals(_referenceModel.hashCode() , _typeNotEquivelentModel.hashCode());
	}
}

File: dotnet/tests/amp.tests/rabbit/topology/ExchangeTests.cs
using System;
using System.Collections.Generic;
using amp.rabbit.topology;
using NUnit.Framework;

namespace amp.tests.rabbit.topology
{
    /// <summary>
    /// These tests assert that equality and hashcode generation use value semantics
    /// not just in-so-far as properties have the same reference values but that 
    /// properties themselves are compared using a value semantic.  
    /// </summary>
    [TestFixture]
    public class ExchangeTests : AmqpBaseModelTests
    {
        protected Exchange _typeNotEquivelentModel;

        public override void SetUp()
        {
            _referenceModel = NewExchange("topic", "arg1");
            _equivelentModel = NewExchange("topic", "arg1");
            _notEquivelentModel = NewExchange("topic", "arg2");
            _typeNotEquivelentModel = NewExchange("direct", "arg1");
        }

        private Exchange NewExchange(string exchangeType, string argValue)
        {
            var exchange = new Exchange("ex1", exchangeType, arguments: new Dictionary<String, object>());
            exchange.Arguments.Add("A", argValue);
            return exchange;
        }

        [Test]
        public void Equality_should_consider_exchange_type()
        {
            Assert.That(_typeNotEquivelentModel.Equals(_referenceModel), Is.False);
        }

        [Test]
        public void Hashcodes_should_differ_if_exchange_type_differ()
        {
            Assert.That(_typeNotEquivelentModel.GetHashCode(), Is.Not.EqualTo(_referenceModel.GetHashCode()));
        }
    }
}
